lib/value/src/kind/accessor.rs

```diff
@@ -0,0 +1,677 @@
+use std::{borrow::Cow, collections::BTreeMap};
+
+use lookup::{Field, Lookup, Segment};
+
+use super::{Collection, Kind};
+
+impl Kind {
+    /// Find the [`Kind`] at the given path.
+    ///
+    /// If the path points to root, then `self` is returned, otherwise `None` is returned if `Kind`
+    /// isn't an object or array. If the path points to a non-existing element in an existing collection,
+    /// then the collection's `unknown` `Kind` variant is returned.
+    ///
+    /// Negative indexing always returns `None`, as the type system doesn't know the exact size of
+    /// an array, and thus cannot count backward.
+    #[must_use]
```

Comment:
I'm not convinced these `must_use` annotations are useful since I can't see any scenario where someone would call this function and forget to use the result. 

I think the intention behind this is for situations where the functions success path returns no useful result, but the error path returns an error. A developer could forget to handle the error and think their code is working fine.

Reply:
The documentation (and RFC) has examples that specifically call out any use-case where the result of a function should be used for that function to be useable (e.g. for a pure function that has no side effects).

See: https://doc.rust-lang.org/reference/attributes/diagnostics.html#the-must_use-attribute

In fact, there's [ongoing work](https://github.com/rust-lang/rust/issues/89692) to expand this attribute across Rust's stdlib as well (such as `len()` and `is_empty()`).

See also this "[for reviewers](https://std-dev-guide.rust-lang.org/code-considerations/design/must-use.html#for-reviewers)" section, which calls out:

> Look for any legitimate use-cases where `#[must_use]` will cause callers to explicitly ignore values. If these are common then `#[must_use]` probably isn't appropriate.

In this (and other) case(s), calling `find_at_path`, but _not_ using the return value is clearly a programmer bug, and so a valid case to apply `#[must_use]` to.

Personally, I don't mind _too much_ whether we add it or not, but it's a nice tool to have in our toolbox to prevent any bugs in API misuse down the line, and while it does add a bit of clutter, I'm mostly on the side that the clutter is worth the extra safety.